source/debian/spybrowse.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH SPYVIEW SECTION "May  6, 2003"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
spybrowse \- interactive viewer launcher
.SH SYNOPSIS
.B spybrowse
.br
.SH DESCRIPTION
.PP
A browser program for interactively loading images from the current
directory into spyview and spyview3d. 
.PP
The pattern field supports standard shell wildcards and pattern
matching (see fl_filename_match in fltk documentation.)
.PP
Clicking on an already selected item will load it into spyview.
.SH SEE ALSO
.BR spyview3d (1),
.BR spyview (1)
.br
.SH AUTHOR
This manual page was written by Gary Steele <gsteele@electron.mit.edu>,
for the Debian project (but may be used by others).







debian/changelog
spyview (0.1-38) local; urgency=low

  * Serious changes since the last update!

 -- gsteele <gsteele@newkavli.nano.tudelft.nl>  Sat, 18 Aug 2007 11:02:30 -0400

spyview (0.1-37) local; urgency=low

  * Added loads of stuff, including the sweet new postscript output, and 
    an updated man page.

 --  <gsteele@qtmartini.sqt.tn.tudelft.nl>  Fri, 24 Nov 2006 16:43:46 +0100

spyview (0.1-36) local; urgency=low

  * Minor update to man page, very significant update to the software...

 --  <gsteele@qtmartini.sqt.tn.tudelft.nl>  Wed, 25 Oct 2006 09:55:37 +0200

spyview (0.1-35) local; urgency=low

  * Too many new features to list...including severely outdated man pages...

 -- odie <gsteele@qtmartini.sqt.tn.tudelft.nl>  Fri, 25 Aug 2006 11:58:50 +0200

spyview (0.1-34) local; urgency=low

  * Added pdf files and a lot of new features.

 -- odie <odie@electron.mit.edu>  Fri, 19 May 2006 05:43:36 -0400

spyview (0.1-33) local; urgency=low

  * Included stminfo in the distribution.

 --  <gsteele@vortex.ashoori.mit.edu>  Mon, 25 Apr 2005 16:27:49 -0400

spyview (0.1-32) local; urgency=low

  * Added support for a 3D binary floating point format that I call "mtx"
  * Added HSV rotation wheel for adjusting colors of a colormap
  * Added a new blue-green-pink colormap designed specifically for 
    my incompressible strip simulation data, as well as a new transfer 
    function adjustment that centers the colormap on the peak of the 
    histogram
  * Added a column-by-column subtraction routine to the image processing.

 --  <gsteele@vortex.ashoori.mit.edu>  Sat, 22 Jan 2005 15:53:15 -0500

spyview (0.1-31) local; urgency=low

  * UI enhancements, added support for arbitrary sized colormaps
  * Parameters are now saved in PPM file as comments
  * Modified matlabpink colormap so that it does not have a pre-applied 
    gamma correction

 --  <gsteele@vortex.ashoori.mit.edu>  Sun, 21 Nov 2004 17:47:10 -0500

spyview (0.1-30) local; urgency=low

  * Improved histogram plot

 --  <gsteele@vortex.ashoori.mit.edu>  Tue,  2 Nov 2004 10:36:04 -0500

spyview (0.1-29) local; urgency=low

  * Big enhancements to spybrowse: added a new Fl_Listbox widget

 --  <gsteele@vortex.ashoori.mit.edu>  Tue,  2 Nov 2004 08:49:57 -0500

spyview (0.1-28) local; urgency=low

  * Added support for separate x and y zooms, updated man page.

 --  <gsteele@vortex.ashoori.mit.edu>  Thu,  5 Aug 2004 16:45:20 -0400

spyview (0.1-27) local; urgency=low

  * Increased line buffer size of dat2pgm to 1024*1024 bytes. Also 
    fixed the logarithm so that it correctly output the base10 exponent.

 --  <gsteele@vortex.ashoori.mit.edu>  Wed,  4 Aug 2004 16:35:58 -0400

spyview (0.1-26) local; urgency=low

  * Removed ugly colormaps again after they snuck back in.

 --  <gsteele@vortex.ashoori.mit.edu>  Tue, 27 Jul 2004 12:52:10 -0400

spyview (0.1-25) local; urgency=low

  * Added a program dat2pgm for converting matrix ascii files into 
    pgm files that include the spyview comment fields for the axis
    scaling.

 --  <gsteele@vortex.ashoori.mit.edu>  Wed, 21 Jul 2004 21:09:09 -0400

spyview (0.1-24) local; urgency=low

  * Added a bunch of stuff, including arbitrary line cuts, support for
    changing x,y,z scaling/units of image using a dialog box, and other
    great stuff. Also, took topometrix colormaps out of distribution, since 
    they were just clogging things up and were mostly crap anyway.

 --  <gsteele@vortex.ashoori.mit.edu>  Fri,  2 Apr 2004 20:23:13 -0500

spyview (0.1-23) local; urgency=low

  * Added more normalize on load options to spyview.

 --  <gsteele@vortex.ashoori.mit.edu>  Tue,  2 Mar 2004 14:54:15 -0500

spyview (0.1-22) local; urgency=low

  * Added some colormaps, finally fixed filename stuff so we can load
    files that are in subdirectories.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 20 Jan 2004 18:30:16 -0500

spyview (0.1-21) local; urgency=low

  * Fixed lbl subtraction.

 -- Gary Steele <gsteele@electron.mit.edu>  Sat, 10 Jan 2004 19:45:21 -0500

spyview (0.1-20) local; urgency=low

  * Added a new program spyrotate 
  * Some code cleanup

 -- Gary Steele <gsteele@vortex.ashoori.mit.edu>  Tue,  9 Dec 2003 14:37:12 -0500

spyview (0.1-19) local; urgency=low

  * Some bugfixes, changes shortcut 's' to save file as ppm. 

 -- Gary Steele <gsteele@vortex.ashoori.mit.edu>  Wed, 26 Nov 2003 10:50:48 -0500

spyview (0.1-18) local; urgency=low

  * Fixed a bug with the plane subtraction in the saveFile routine.

 -- Gary Steele <gsteele@vortex.ashoori.mit.edu>  Sun, 19 Oct 2003 13:33:22 -0400

spyview (0.1-17) local; urgency=low

  * Added an shlib.local file to override libc dependecy.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue,  2 Sep 2003 15:15:50 -0400

spyview (0.1-16) local; urgency=low

  * Recompiled with gcc-3.3

 -- Gary Steele <gsteele@electron.mit.edu>  Tue,  2 Sep 2003 12:38:16 -0400

spyview (0.1-15) local; urgency=low

  * Fixed up gl2ps scaling, added rollers for colormap in spyview3d

 -- Gary Steele <gsteele@electron.mit.edu>  Sun,  3 Aug 2003 17:04:58 -0400

spyview (0.1-14) local; urgency=low

  * Changed gnuplot xsection commands so you can zoom with the mouse
    properly.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 29 Jul 2003 15:09:35 -0400

spyview (0.1-13) local; urgency=low

  * Added plane subtraction to spyview3d. 

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 22 Jul 2003 22:29:31 -0400

spyview (0.1-12) local; urgency=low

  * Added basic colormap manipulation support to spyview3d.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 22 Jul 2003 16:16:21 -0400

spyview (0.1-11) local; urgency=low

  * Added support for reading unit and scales out of AFM and Stm 
    files. The crosssections are now plots showing physical units.
    (The zaxis unit is ignored for Stm files, since it's usually
    still +/- 10V).
    
 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 22 Jul 2003 14:15:09 -0400

spyview (0.1-10) local; urgency=low


source/spyview/spybrowse.C
#include <stdio.h>
#include "spybrowse_ui.h"
#include "spybrowse.h"
#include "message.h"
#include <Fl/filename.H>
#include <Fl/fl_ask.H>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>

using namespace std;

string spyview_cmd;

int main(int argc, char **argv)
{
  spyview_cmd="spyview ";

  Fl::visual(FL_RGB);
  Fl::get_system_colors();
  
  // If we do not have spyview in the default path, we need to give a
  // full path to spyview.

  // The only time that spyview is not in the default path is if it is
  // being run under win32 not from the command line.

  // In this case, argv[0] will contain the full path to where
  // spybrowse was executed, and thus, also to where the spyview
  // executable is.

#ifdef WIN32
  if (getenv("SPYVIEW_DIR") != NULL) // otherwise get a windows crash?
    spyview_cmd = getenv("SPYVIEW_DIR");
  else
    {
      // Spyview is not in the path...probably some dumb windows
      // user...fortunately, they probably clicked on the
      // spybrowse.exe, so we can get the path from argv[0];
      spyview_cmd = argv[0];
      int c2 = spyview_cmd.find_last_of("\\");
      if(c2 >= 0)
	spyview_cmd.replace(c2,string::npos, "");
      spyview_cmd += "\\";
      info("spyview not in path: guessing %s as path from argv[0]\n", spyview_cmd.c_str());
    }
  if (spyview_cmd.size() == 0)
    {
      warn("Could not find good spyview_cmd: reverting to '.'\n");
      spyview_cmd = ".";
    }
  spyview_cmd = "start " + spyview_cmd + "spyview_console.exe ";
  info("spyview_cmd is %s\n", spyview_cmd.c_str());
#endif

  make_window();
  update_pattern();
  patternbox->value("*.{Stm,pgm,TFR,TRR,FRR,FFR,dat,mtx}");
  string title = "spybrowse ";
  char buf[1024*4]; 
  if (getcwd(buf,1024) == NULL)
    {
      fprintf(stderr, "directory longer than 4kbytes?");
      exit(-1);
    }
  directory->value(buf);
  title += directory->value();
  refreshb->do_callback();
  win->label(title.c_str());
  win->show();
  Fl::run();
}

void update_pattern()
{
  string pattern = "*.{";
  if (dat->value())
    pattern += "dat,";
  if (mtx->value())
    pattern += "mtx,";
  if (pgm->value())
    pattern += "pgm,";
  if (other->value())
    pattern += "Stm,TFR,TRR,FRR,FFR";
  pattern += "}";
  patternbox->value(pattern.c_str());
  refreshb->do_callback();
}

 
void refreshb_cb(Fl_Widget *, void *)
{
  dirent **list;
  int n;
  
  n = fl_filename_list(directory->value(), &list);
  
  fb->clear();
  
  for (int i = 0; i<n; i++)
    if (fl_filename_match(list[i]->d_name, patternbox->value()))
      fb->add(list[i]->d_name);
}

void fb_cb(Fl_Widget *, void*)
{
  int nselected = 0;
  for (int i=1; i<=fb->size(); i++)
    {
      if (fb->selected(i))
	nselected++;
    }
}

void viewb_cb(Fl_Widget *, void*)
{
  int n = 0;
  string cmd = spyview_cmd;
  for (int i = 1; i<=fb->size(); i++)
    if (fb->selected(i))
      {
	cmd += fb->text(i);
	cmd += " ";
	n++;
      }
  cmd += "&";
  if (n == 0)
    fl_alert("No files selected.");
  else
    system(cmd.c_str());
}

void view2b_cb(Fl_Widget *, void*)
{
  int n = 0;
  string cmd = spyview_cmd;
  string tmp;
  for (int i = 1; i<=fb->size(); i++)
    if (fb->selected(i))
      {
	tmp = cmd;
	tmp += fb->text(i);
	tmp += "&";
	info("command %s\n", tmp.c_str());
	system(tmp.c_str());
#ifndef WIN32
	sleep(1);
#else
	Sleep(1000);
#endif
      }
}
